### ISSUE TITLE
Passing 3 or more lists to _.zipWith doesn't make any sense

### ISSUE THREAD

**Issue by tullmann** [lodash/lodash#1484]
It looks like the iteratee function is being invoked for each pair(?) of elements:
```
_.zipWith([10,20,30],[40,50,60],[70,80,90], function() { console.log(Array.prototype.slice.call(arguments)); return -1; })
[ 10, 40, 1, [ 10, 40, 70 ] ]
[ -1, 70, 2, [ 10, 40, 70 ] ]
[ 20, 50, 1, [ 20, 50, 80 ] ]
[ -1, 80, 2, [ 20, 50, 80 ] ]
[ 30, 60, 1, [ 30, 60, 90 ] ]
[ -1, 90, 2, [ 30, 60, 90 ] ]
[ -1, -1, -1 ]
```

According to the documentation the callback should be invoked with "(accumulator, value, index, group)", but you can see the accumulator only comes back on every other invocation.  And the index bounces between 1 and 2, which seems wrong.

My expectation is that the `zipWith` iteratee function would be invoked with one argument from each list, and nothing more.  (You can make a case to maybe add an index at the end).  So I'd expect to see something more like:
```
_.zipWith([10,20,30],[40,50,60],[70,80,90], function() { console.log(Array.prototype.slice.call(arguments)); return -1; })
[ 10, 40, 70]
[ 20, 50, 80]
[ 30, 60, 90]
```

I think the `_.add` example in the doc is a bit too forgiving.  Something with string concatenation would probably be a better example and test.  E.g., this is what I get from lodash 3.9.1:
```
lodash.zipWith(['a','b','c'],['x','y','z'],['1','2','3'], function(a1, a2, a3) { return a1 + a2 + a3; })
[ 'ax112', 'by122', 'cz132' ]
```
which does not, AFAICT, make any sense.

But maybe I'm misreading what/how zipWith is supposed to work and the current semantics make sense, but then the documentation needs a lot of work.

**Comment by jdalton** [comment]
Because of how reducing works the first item in the group is used as the initial accumulator value which is why the index goes from `1` then `2`. I'll add a doc note for it this evening.

**Comment by TurplePurtle** [comment]
Is there a reason for using zipWith in the current way? The way @tullmann mentioned makes far more sense, and makes things far less verbose IMO.

Consider joining 3 arrays `xs`, `ys`, `zs` into an array of objects `objs`, where each object has fields `{ x, y, z }`.

In the existing implementation, this would be done as:

```javascript
var objs = _.zipWith(xs, ys, zs, (acc, value, index, group) => {
  if (index === 1) {
    return {x: acc, y: value};
  } else {
    acc.z = value;
    return acc;
  }
});
```

However, doing it in a way similar to how it is done in Haskell or RxJS this becomes much simpler:

```javascript
var objs = zipWith(xs, ys, zs, (x, y, z) => ({ x, y, z }));
```

Perhaps I can see a reason for using the first way if you are building a very large object, but having to branch on the first index anytime you want to zipWith 3 or more arrays seems excessive to me in the typical use case.

(Here is a reference implementation of the second method https://gist.github.com/TurplePurtle/a0a9448d41ea5a6b6998 )

**Comment by jdalton** [comment]
I have no heart burn changing the behavior. Let's slip this behavior change in for v4.

**Comment by jdalton** [comment]
Patched https://github.com/lodash/lodash/commit/6a964fdc57cc291da73689b3bb6dac6ad2649f6f.

**Comment by TurplePurtle** [comment]
I attempted to patch it myself, but noticed that using `apply(iteratee, undefined, group)` did not respect thisArg (at least in my testing). Is that an issue here?

**Comment by jdalton** [comment]
In v4 there is no `thisArg` in favor of folks using `_.bind`.

**Comment by rafaelbiten** [comment]
I'd like to ask your opinion on something that somewhat related to this:

Today, while using `zipWith` in a scenario where order matters, like:
```
zipWith(xs, ys, zs, (x, y, z) => ({ x, y, z }))
```

I noticed some inconsistencies in the results. In some cases, I noticed some `zs` appearing in a place where there should only be `ys`. 

It took me a while to understand what was going on and the reason of the `unexpected` behaviour  was that in some cases `ys` was `undefined` (not even an empty `[]`).

Knowing that, I fixed it by guarding each value passed to `zipWith` with `|| []` and that fixed the issue.

Now, my question is: Isn't there a better way to handle this scenario?

I just checked with `Ramda` and on that scenario, it will simply `throw`, which would have helped me understand what was going on much faster.

Another option would be to default `undefined` to an `[]` internally, but I guess that could end up hiding bugs.

Thank you!
